### ISSUE TITLE
How to add new key binding to neotree buffer

### ISSUE THREAD

**Issue by saga92** [syl20bnr/spacemacs#4042]
I would like to add key `o` to enter the folder or file, but not work when add the code 
```lisp
(define-key neotree-mode-map (kbd "o") 'neotree-enter) 
```
into dotspacemacs

**Comment by StreakyCobra** [comment]
Try this in `dotspacemacs/user-config`:

```
(with-eval-after-load 'neotree
  (add-hook 'neotree-mode-hook
            (lambda ()
              (define-key evil-motion-state-local-map (kbd "o") 'neotree-enter))))
```

**Comment by bmag** [comment]
While @StreakyCobra's solution should work, it's a bit extreme, because it bind the key every time neotree buffer is created (although that buffer should only be created once, so it's not such a big deal). Adding the following to my `dotspacemacs/user-config` was sufficient for me:
```elisp
(with-eval-after-load 'neotree
  (define-key neotree-mode-map (kbd "o") 'neotree-enter))
```

**Comment by StreakyCobra** [comment]
@bmag You're right. For `o` it will work :-)

FYI if you want to remap on of these letters:
```
        (define-key evil-motion-state-local-map (kbd "TAB")  'neotree-stretch-toggle)
        (define-key evil-motion-state-local-map (kbd "RET")  'neotree-enter)
        (define-key evil-motion-state-local-map (kbd "|")    'neotree-enter-vertical-split)
        (define-key evil-motion-state-local-map (kbd "-")    'neotree-enter-horizontal-split)
        (define-key evil-motion-state-local-map (kbd "?")    'evil-search-backward)
        (define-key evil-motion-state-local-map (kbd "c")    'neotree-create-node)
        (define-key evil-motion-state-local-map (kbd "d")    'neotree-delete-node)
        (define-key evil-motion-state-local-map (kbd "gr")   'neotree-refresh)
        (define-key evil-motion-state-local-map (kbd "h")    'spacemacs/neotree-collapse-or-up)
        (define-key evil-motion-state-local-map (kbd "H")    'neotree-select-previous-sibling-node)
        (define-key evil-motion-state-local-map (kbd "J")    'neotree-select-down-node)
        (define-key evil-motion-state-local-map (kbd "K")    'neotree-select-up-node)
        (define-key evil-motion-state-local-map (kbd "l")    'spacemacs/neotree-expand-or-open)
        (define-key evil-motion-state-local-map (kbd "L")    'neotree-select-next-sibling-node)
        (define-key evil-motion-state-local-map (kbd "q")    'neotree-hide)
        (define-key evil-motion-state-local-map (kbd "r")    'neotree-rename-node)
        (define-key evil-motion-state-local-map (kbd "R")    'neotree-change-root)
        (define-key evil-motion-state-local-map (kbd "s")    'neotree-hidden-file-toggle))
```

You will have to use my code, because those get recreated at [each time buffer is entered](https://github.com/syl20bnr/spacemacs/blob/develop/layers/+distribution/spacemacs/packages.el#L1574-L1600) :-/

**Comment by bmag** [comment]
Oh my, why did we implement it like this? Is there any reason not to bind the keys normally?

**Comment by StreakyCobra** [comment]
No idea why. If you're motivated to try `evilifying` it, you're welcome :relaxed:

**Comment by bmag** [comment]
Will try to find the time :wink:
By the way, related issues: #3752 #2783

**Comment by justbur** [comment]
The local state maps are buffer local so I think you have to do it this way to use those. That's also what evil leader did by the way. 

I initially set it up to use the global state maps but people kept complaining about keys that weren't available in certain buffers (essentially because some modes were intentionally overriding the global state maps). Using the local state maps fixes this in a overly strong manner imo but that's what people seemed to want (parity with evil leader).

**Comment by justbur** [comment]
It's also not each time the buffer is entered its when the buffer is created and evil mode is turned on.

**Comment by justbur** [comment]
I'm sorry I should have looked at the link. I was describing how bind-map works now  not how neotree key bindings are done. But the concept turns out to be similar actually.

**Comment by saga92** [comment]
@StreakyCobra Thanks a lot:-) It works when add your code into `dotspacemacs/user-config`, however I try to add more key binding such as 
```lisp
(with-eval-after-load 'neotree
    (add-hook 'neotree-mode-hook
              (lambda()
                (define-key evil-motion-state-local-map (kbd "o") 'neotree-enter)
                (define-key evil-motion-state-local-map (kbd "i") (neotree-make-executor
                                                                   :file-fn 'neo-open-file-horizontal-split))
                (define-key evil-motion-state-local-map (kbd "s") (neotree-make-executor
                                                                   :file-fn 'neo-open-file-vertical-split))
                (define-key evil-motion-state-local-map (kbd "S") 'neotree-hidden-file-toggle)
                )))
```
`s` can't open file in vertical split, it will still perform as before to hidden the file toggle.

**Comment by StreakyCobra** [comment]
You can't use:
```
(neotree-make-executor :file-fn 'neo-open-file-horizontal-split)
```
you have to use syntax like:
```elisp
(lambda () (neotree-make-executor :file-fn 'neo-open-file-horizontal-split))
```

Same for the line after.

**Comment by saga92** [comment]
it reports `wrong type argument` when using
```lisp
(lambda () (neotree-make-executor :file-fn 'neo-open-file-horizontal-split))
```

**Comment by StreakyCobra** [comment]
Oh, I didn't really looked at it. Let me check carefully then :-)

**Comment by StreakyCobra** [comment]
indeed `(neotree-make-executor)` generates a lambda function already.

**Comment by StreakyCobra** [comment]
My bad, I forgot to copy one parameter in my first snippet:

```elisp
  (with-eval-after-load 'neotree
    (add-hook 'neotree-mode-hook
              (lambda()
                (define-key evil-motion-state-local-map (kbd "o") 'neotree-enter)
                (define-key evil-motion-state-local-map (kbd "i") (neotree-make-executor
                                                                   :file-fn 'neo-open-file-horizontal-split))
                (define-key evil-motion-state-local-map (kbd "s") (neotree-make-executor
                                                                   :file-fn 'neo-open-file-vertical-split))
                (define-key evil-motion-state-local-map (kbd "S") 'neotree-hidden-file-toggle))
              'append))
```

The `'append` at the end.

**Comment by saga92** [comment]
@StreakyCobra It works perfect! Sincerely appreciate for your help:-)

**Comment by StreakyCobra** [comment]
My pleasure. If you think your question has been answered and there is nothing more to do on this issue, you can consider closing it :relaxed:

**Comment by saga92** [comment]
@StreakyCobra Got it :+1:

**Comment by bmag** [comment]
@saga92 @StreakyCobra just a heads up, I opened a PR to evilify Neotree, which will probably nullify the solution and require a new one.

**Comment by syl20bnr** [comment]
This is merged in develop, I guess that binding directly into the neotree keymap should work.

**Comment by kmelva** [comment]
I have the same problem as @saga92 - I want to rebind movement keys in NeoTree, let's say `i` for `neotree-previous-line` and `k` for `neotree-next-line`. I've tried a lot of things but nothing works, this should work based on all my other key bindings:

```
(with-eval-after-load 'neotree
  (evil-define-key 'normal neotree-mode-map (kbd "i") 'neotree-previous-line)
  (evil-define-key 'normal neotree-mode-map (kbd "k") 'neotree-next-line))
```

but it doesn't do anything. If I press `i` I get `i is undefined` in minibuffer, and `k` moves the cursor up.

Any ideas what should be the Spacemas way of doing this? I have similar bindings for `dired-mode-map` and it works just fine.

**Comment by bmag** [comment]
@kmelva neotree buffer has been evilified, so now it opens in `evilified` state. Try binding in `evilified` state instead of `normal` state. You might want to read up on "evilified state" in Spacemacs docs and on how key bindings work. If you need more help, you can open a new issue report.

**Comment by kmelva** [comment]
@bmag thanks, that's it, working great. I made a classical mistake of not RTFMing deep enough. I've read this issue, and when I saw NeoTree was `evilified` I just thought that means I have to use `evil-define-key` instead of `define-key` and it will work. Thanks again! 

I did ask on [StackOverflow](http://stackoverflow.com/questions/41314759/rebind-keys-in-spacemacs-for-neotree-buffer) but no one responded. If you want some SO points paste your answer there and I'll mark it as answered :).
